nomad/stream/ndjson.go

```diff
@@ -0,0 +1,93 @@
+package stream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var (
+	// JsonHeartbeat is an empty JSON object to send as a heartbeat
+	// Avoids creating many heartbeat instances
+	JsonHeartbeat = &structs.EventJson{Data: []byte("{}")}
+)
+
+// JsonStream is used to send new line delimited JSON and heartbeats
+// to a destination (out channel)
+type JsonStream struct {
+	// mu locks during send and protects running
+	mu sync.Mutex
+
+	// ctx is a passed in context used to notify the json stream
+	// when it should terminate
+	ctx context.Context
+
+	outCh chan *structs.EventJson
+
+	// heartbeat is the interval to send heartbeat messages to keep a connection
+	// open.
+	heartbeatTick *time.Ticker
+}
+
+// NewJsonStream creates a new json stream that will output Json structs
+// to the passed output channel. The constructor starts a goroutine
+// to begin heartbeating on its set interval.
+func NewJsonStream(ctx context.Context, heartbeat time.Duration) *JsonStream {
+	s := &JsonStream{
+		ctx:           ctx,
+		outCh:         make(chan *structs.EventJson, 10),
+		heartbeatTick: time.NewTicker(heartbeat),
+	}
+
+	go s.heartbeat()
+
+	return s
+}
+
+func (n *JsonStream) OutCh() chan *structs.EventJson {
+	return n.outCh
+}
+
+func (n *JsonStream) heartbeat() {
+	for {
+		select {
+		case <-n.ctx.Done():
+			return
+		case <-n.heartbeatTick.C:
+			// Send a heartbeat frame
+			select {
+			case n.outCh <- JsonHeartbeat:
+			case <-n.ctx.Done():
+				return
+			}
+		}
+	}
+}
+
+// Send encodes an object into Newline delimited json. An error is returned
+// if json encoding fails or if the stream is no longer running.
+func (n *JsonStream) Send(v interface{}) error {
```

Comment:
I think you want to prevent spurious heartbeats every time you send by doing something like:

```go
n.heartbeatTick.Stop()
defer n.heartbeatTick.Reset(n.heartbeat)
```

Reply:
fwiw - i always conflicted about this point here and elsewhere.  As we heartbeat every 30 seconds, spurious heartbeats isn't really an issue. When considering the overhead of resetting ticker on every message (for busy cluster that's a lot), it's not a clear win to reset on every message.

Reply:
I originally had this when the heartbeat was only 3-5 seconds, after doing some research and reading an article @DingoEatingFuzz pointed out, a much greater heartbeat of 30 seconds seemed fine and not worth resetting on

Reply:
I would expect `timer.{Stop,Reset}` to be far cheaper than a chan send + the associated network IO. The implementation even hints network IO uses a timer itself: https://github.com/golang/go/blob/681559e1f10f83a053b4ebab101de3d77ede8353/src/runtime/time.go#L405-L407

That being said this is definitely not a blocker. Once every 30s isn't likely to impact anything until there are hundreds of consumers or more.